src/Transform/ONNX/SimplifyShapeRelatedOps.cpp

```diff
@@ -0,0 +1,519 @@
+bool areDimsFromConcat(Value val) {
+  return (areDims(val) && definedBy<ONNXConcatOp>(val));
+}
+
+/// Get all dimensions that are stored by the value.
+void getDims(Value val, SmallVectorImpl<Value> &dims) {
+  assert(areDims(val) && "Value does not store dimensions");
+  if (definedBy<ONNXConcatOp>(val)) {
+    for (Value v : val.getDefiningOp()->getOperands()) {
+      SmallVector<Value, 4> inputs;
+      getDims(v, inputs);
+      for (Value i : inputs)
+        dims.emplace_back(i);
+    }
+  } else
+    dims.emplace_back(val);
+}
+
+/// Get all dimensions in I64 (-1 for unknown) that are stored by the value.
+void getDimsInt64(Value val, SmallVectorImpl<int64_t> &result) {
+  SmallVector<Value, 4> dims;
+  getDims(val, dims);
+  for (Value v : dims) {
+    if (auto constOp = dyn_cast<ONNXConstantOp>(v.getDefiningOp())) {
+      auto valueAttr = constOp.valueAttr().cast<DenseElementsAttr>();
+      int64_t dim = valueAttr.getSplatValue<int64_t>();
+      result.emplace_back(dim);
+    } else {
+      result.emplace_back(-1);
+    }
+  }
+}
+
+/// Create a ConcatOp to concat the list of tensors.
+Value emitConcatOpForDims(
+    MultiDialectBuilder<OnnxBuilder> create, ValueRange inputs) {
+  int64_t rank = inputs.size();
+  if (rank == 1)
+    return inputs[0];
+  Type elementType = getElementType(inputs[0].getType());
+  Type resultType = RankedTensorType::get({rank}, elementType);
+  Value concatOutput = create.onnx.concat(resultType, inputs, /*axis=*/0);
+  return concatOutput;
+}
+
+/// Update the function signature if the op's output is the return value.
+void updateFunctionSignature(Operation *op) {
+  assert(op->getResults().size() == 1 && "Only support single result ops");
+  Operation *parentOp = op->getParentOp();
+  if (auto f = dyn_cast<func::FuncOp>(parentOp)) {
+    if (!f.back().empty() && f.back().back().hasTrait<OpTrait::IsTerminator>())
+      if (auto terminator = f.getBody().back().getTerminator()) {
+        for (Operation *user : op->getResults()[0].getUsers()) {
+          if (user == terminator) {
+            auto results = user->getOperandTypes();
+            f.setType(FunctionType::get(f.getContext(),
+                f.getFunctionType().getInputs(),
+                std::vector<Type>(results.begin(), results.end())));
+          }
+        }
+      }
+  }
+}
+
+/// Rewrite onnx.Shape into onnx.Dim and onnx.Concat.
+//
+/// For example, the folowing onnx.Shape:
+/// %0 = "onnx.Shape"(%arg0) : (tensor<?x256xi64>) -> tensor<2xi64>
+///
+/// will be rewritten into:
+///
+/// %0 = "onnx.Dim"(%arg0) {axis = 0 : si64} :
+///           (tensor<?x256xi64>) -> tensor<1xi64>
+/// %1 = "onnx.Constant"() {value = dense<256> : tensor<1xi64>} :
+///           () -> tensor<1xi64>
+/// %2 = "onnx.Concat"(%0, %1) {axis = 0 : si64, for_dims = true} :
+///           (tensor<1xi64>, tensor<1xi64>) -> tensor<2xi64>
+class ShapeToDimConcatPattern : public OpRewritePattern<ONNXShapeOp> {
```

Comment:
PR #1684 is ready to land, and it adds the opset 15 support for start and end. Do you mind waiting for that PR to land and add support for this here? It would be much easier for you to do, as opposed for @philass to add it if yours land first. Thanks

Reply:
Yes, it's absolutely fine. I'll wait for #1684.

Reply:
Thanks @tungld! I didn't realize you guys were waiting for me. Much appreciated

I'll fix the CI failures in my PR right away, and merge it into head once thats done.

Reply:
Thanks

Reply:
@AlexandreEichenberger since #1684 was landed, I updated this patch to support ShapeOp opset 15.

@philass I found that the lowering of ShapeOp opset 15 failed if start is negative. It does not affect this patch, but I created #1710 to report that issue. Could you take a look the issue? Thanks.